### DISCUSSION TITLE
Request for range(d) airstreams

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/airdrops#29] [Ideas]
A user requested the following feature for airstreams: a synchronized end time for all streams claimed from an airstream campaign. 

This would involve setting a specific end time for all streams, such that all claimed streams end on the same date. It basically results in support for `createWithRange` alongside the current `duration`-driven claim function.

Some questions:
- What should the start time be set to? The claim date (meaning streams will end up with different durations) or a hard value provided by the admin at campaign creation?
- Should the end time have any connection to the airstream's expiry date?

<img width="457" alt="Screenshot 2024-02-05 at 12 19 05" src="https://github.com/sablier-labs/v2-periphery/assets/15740806/9a4a6be4-0f8c-41bb-b2c4-f12b7ca89bb1">

**Comment by PaulRBerg** [comment]
Thanks for this. Feel free to use the `@sablier-labs/@engineers` group or tag people directly so that these discussions end up in our email inboxes.

**Comment by PaulRBerg** [comment]
Interesting proposal. I see why it can be useful for users.

> What should the start time be set to?
> ...
> (i) The claim date (meaning streams will end up with different durations) or (ii) a hard value provided by the admin at campaign creation?

The latter seems more desirable. I have a hunch that that's what 5ire wants (though they should confirm).

1. Different start times would mean different streaming rates.
2. Hard-coded start times would mean identical vesting periods even when the airstream is not claimed.

> Should the end time have any connection to the airstream's expiry date?

In most cases, they would have the same value, but they should still be maintained as separate variables for flexibility purposes.

**Reply by smol-ninja** [reply]
> In most cases, they would have the same value

Why would the expiry date have the same value as the end time?

One identifies whether the airstream can be claimed and other identifies the end date of a claimed stream. So, I think, in most cases, expiry date would be less than the end time (longer vesting but shorter claim window).

**Reply by PaulRBerg** [reply]
Good point. I think we need to ask users what they want.

**Comment by PaulRBerg** [comment]
Looks like they (5ire) want the start time to be the claim time:

<img width="500" alt="SCR-20240207-ktmd" src="https://github.com/sablier-labs/v2-periphery/assets/8782666/4c75e1b8-adfd-4623-88f6-9a54334ed575">

**Reply by smol-ninja** [reply]
Wow. So that means different streaming rates for each claim.

**Reply by PaulRBerg** [reply]
Yep, though there's nothing wrong with that

**Reply by adamgall** [reply]
For what it's worth, the team at @decentdao have this exact same feature request :)

**Comment by PaulRBerg** [comment]
Another reason to implement ranged airstreams: [monthly unlocks for airstreams](https://github.com/sablier-labs/v2-periphery/discussions/312).

**Comment by smol-ninja** [comment]
Request from Arcana for Airstreams ending on the same date:

<img width="495" alt="Screenshot 2024-05-02 at 18 11 50" src="https://github.com/sablier-labs/v2-periphery/assets/6676622/a7003e97-d0d3-44d1-905b-fa278aad2138">

**Reply by PaulRBerg** [reply]
Kinda obvious at this point that the next version of Lockup will include ranged airstreams

**Comment by smol-ninja** [comment]
I have started a discussion that can enable range(d) airstream
- https://github.com/sablier-labs/v2-core/discussions/997

**Comment by PaulRBerg** [comment]
Just received a request for this feature from Protocol Guild, a potentially large user.

<img width="400" alt="SCR-20241009-mmvd" src="https://github.com/user-attachments/assets/351c7910-e10f-4454-b16c-5e54adb62c29">

@smol-ninja could you refresh my mind on this, please? I gather that we have already implemented this feature through the following PRs:

- https://github.com/sablier-labs/v2-core/issues/982
- https://github.com/sablier-labs/v2-core/pull/1024

Right?

CC @sablier-labs/frontend since this will have to be implemented in the UI, too

**Reply by smol-ninja** [reply]
Yes thats right. Here is the discussion: https://github.com/sablier-labs/v2-core/discussions/997. `createWithTimestamps` in Airstreams enables ranged streams whereas allowing end time to be in the past enables instant airdrops (without even to have 1 second stream like how we are doing it rn).

**Comment by adamgall** [comment]
The team at @decentdao has this same feature request. We would love to be able to create Airstreams in which the "start time" of each stream is the claim moment, and the end time for each stream is hardcoded at Airstream creation and the same for each of the Airstreams.

**Reply by PaulRBerg** [reply]
Hi @adamgall, thanks for sharing your feedback.

Your feature request is somewhat different from what we had envisioned in this thread.

In the next release of the protocol (in Q1 2025), we will make it possible to have a fixed start time and a fixed end time.

Even if users claim later than the start time, there will be a portion that will be immediately unlocked upon claiming the airstream.

Is that bad? If yes, could you please share more details about your use case?

**Reply by adamgall** [reply]
My reading comprehension needs some work! Thanks for replying and clarifying ❤️ 

To specifically answer your question:

> Is that bad?

It's not what we're looking for in this case.

Our use case is, admittedly, very non-standard. We are looking for a way for an operator to "queue up" a stream with a specified end time, and a dynamic start time which is set to `block.timestamp` at the moment when the recipient "performs some action" (like claiming the Airstream) onchain.

This would result in an Airstream which all claims have the same end time and (likely) different start times, leading to different streaming rates for each claim in the Airstream.

Another important point about our use case is that we will only have a single claim per Airstream. So really, I think I was just trying to make Airstreams fit into the concept of "recipient can begin the stream at their own discretion". And it does get really close.

But to be very clear, the behavior we're looking for specifically is:
- Operator can "queue up" a stream
  - accepts standard stream params: `sender`, `asset`, `broker`, `totalAmount`, `cancelable`, `transferable`)
  - For `timestamps`, I want to only supply the end date timestamp
- Recipient can, at some point before the end date timestamp, call some onchain function to trigger the "start" of this stream, at which point the start date timestamp is set to that moment, and the stream begins

**Reply by PaulRBerg** [reply]
You're welcome — this was not the clearest discussion, so your reading comprehension wasn't the problem.

It sounds like you might need a different product from what we currently offer. Airstreams are designed for large airdrop campaigns with multiple recipients. However, if I understand correctly, you're looking for a solution where senders can create individual streams with undefined start times until the recipient takes action.

We will consider your feature request. Thank you for sharing it.

**Reply by PaulRBerg** [reply]
I suppose that you want something like this, don't you?

- First chart: user who claims on day 1
- Second chart: user who claims on day 8

<img width="400" src="https://github.com/user-attachments/assets/c00a91ab-f7ff-4a11-8276-5e78d417a2bc"/>
<img width="400" src="https://github.com/user-attachments/assets/3c66d57a-91e7-4c30-8ef2-ed85fbe2a498"/>

**Reply by adamgall** [reply]
Yes @PaulRBerg that's what I'm looking for! It's likely relatively trivial for me to create an intermediate smart contract to support this behavior myself, using your standard Streams (not Airstreams) product.

I'll take that approach for now.

Thanks!

**Reply by PaulRBerg** [reply]
Thanks for confirming!

And glad to hear that an intermediary contract works for now. Will it be open-source? If yes, could you maybe link it here once it's done?

Thank you!

P.S. here's a relevant page from our docs: [Configure Your Local Environment](https://docs.sablier.com/contracts/v2/guides/local-environment)

**Reply by adamgall** [reply]
I've actually taken this conversation back to my team, and we have some more product requirements to work through. If I do end up needing and implementing this contract, I'll definitely share here.

**Reply by PaulRBerg** [reply]
Thanks for the update. You are welcome to ping Max on Telegram (handle: `maxdesalle`) if you would like to brainstorm the implementation.

**Comment by razgraf** [comment]
We've gotten another user request for drops where vesting starts at the same time for everyone.

<img width="441" alt="Screenshot 2024-11-13 at 19 31 39" src="https://github.com/user-attachments/assets/280e4db0-9613-433f-bebf-8b913e391178">

**Comment by PaulRBerg** [comment]
Another request.

<img width="400" alt="SCR-20241223-qdjx" src="https://github.com/user-attachments/assets/f70e3a3f-fdc8-4191-ad00-9f9288d457b0" />

**Comment by PaulRBerg** [comment]
Closing as this has been implemented already.

**Comment by PaulRBerg** [comment]
Worth noting that Nicholas from Seedclub/ $ENJOY passed on using Sablier despite wanting to use it because of the lack of ranged airstreams.

<img width="523" alt="SCR-20241223-tjwj" src="https://github.com/user-attachments/assets/71db4ec8-be06-448d-9297-47d100a9c0c1" />

**Comment by smol-ninja** [comment]
Metaverse HQ is interested into ranged airstreams. A reminder to update them when its live.

<img width="400" alt="Screenshot 2025-01-04 at 11 30 43" src="https://github.com/user-attachments/assets/8918cfa3-acf2-456f-b973-89f15dd17c04" />
